File: src/compat/lib/PaywallPresentationRequestStatus.ts
import { Experiment } from "./Experiment"

/**
 * @category Enums
 * @since 0.0.15
 * The status types for a paywall presentation request.
 */
export enum PaywallPresentationRequestStatusType {
  presentation = "presentation",
  noPresentation = "noPresentation",
  timeout = "timeout",
}

/**
 * @category Models
 * @since 0.0.15
 * Represents the status of a paywall presentation request.
 */
export class PaywallPresentationRequestStatus {
  type: PaywallPresentationRequestStatusType

  private constructor(type: PaywallPresentationRequestStatusType) {
    this.type = type
  }

  static fromJson(json: { [key: string]: any }): PaywallPresentationRequestStatus {
    switch (json.status) {
      case "presentation":
        return new PaywallPresentationRequestStatus(
          PaywallPresentationRequestStatusType.presentation,
        )
      case "noPresentation":
        return new PaywallPresentationRequestStatus(
          PaywallPresentationRequestStatusType.noPresentation,
        )
      case "timeout":
        return new PaywallPresentationRequestStatus(PaywallPresentationRequestStatusType.timeout)
      default:
        throw new Error("Invalid PaywallPresentationRequestStatus type")
    }
  }
}

/**
 * @category Enums
 * @since 0.0.15
 * The reason types for a paywall presentation request status.
 */
export enum PaywallPresentationRequestStatusReasonType {
  debuggerPresented = "debuggerPresented",
  paywallAlreadyPresented = "paywallAlreadyPresented",
  userIsSubscribed = "userIsSubscribed",
  holdout = "holdout",
  noAudienceMatch = "noAudienceMatch",
  placementNotFound = "placementNotFound",
  noPaywallViewController = "noPaywallViewController",
  noPresenter = "noPresenter",
  noConfig = "noConfig",
  subscriptionStatusTimeout = "subscriptionStatusTimeout",
}

/**
 * @category Models
 * @since 0.0.15
 * Represents the reason for a paywall presentation request status.
 */
export class PaywallPresentationRequestStatusReason {
  type: PaywallPresentationRequestStatusReasonType
  experiment?: Experiment

  private constructor(type: PaywallPresentationRequestStatusReasonType, experiment?: Experiment) {
    this.type = type
    this.experiment = experiment
  }

  static fromJson(json: { [key: string]: any }): PaywallPresentationRequestStatusReason {
    switch (json.reason) {
      case "debuggerPresented":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.debuggerPresented,
        )
      case "paywallAlreadyPresented":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.paywallAlreadyPresented,
        )
      case "userIsSubscribed":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.userIsSubscribed,
        )
      case "holdout":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.holdout,
          Experiment.fromJson(json.experiment),
        )
      case "noAudienceMatch":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noAudienceMatch,
        )
      case "placementNotFound":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.placementNotFound,
        )
      case "noPaywallViewController":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noPaywallViewController,
        )
      case "noPresenter":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noPresenter,
        )
      case "noConfig":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noConfig,
        )
      case "subscriptionStatusTimeout":
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.subscriptionStatusTimeout,
        )
      default:
        throw new Error("Invalid PaywallPresentationRequestStatusReason type")
    }
  }
}

File: android/src/main/java/expo/modules/superwallexpo/json/PaywallPresentationRequestStatus.kt
package expo.modules.superwallexpo.json

import com.superwall.sdk.paywall.presentation.internal.PaywallPresentationRequestStatus
import com.superwall.sdk.paywall.presentation.internal.PaywallPresentationRequestStatusReason

fun PaywallPresentationRequestStatus.toJson(): Map<String, Any> {
  val map = mutableMapOf<String, Any>()
  val statusValue = when (this) {
    PaywallPresentationRequestStatus.Presentation -> "presentation"
    PaywallPresentationRequestStatus.NoPresentation -> "noPresentation"
    PaywallPresentationRequestStatus.Timeout -> "timeout"
  }
  map["status"] = statusValue
  return map
}

fun PaywallPresentationRequestStatusReason.toJson(): Map<String, Any> {
  val map = mutableMapOf<String, Any>()
  when (this) {
    is PaywallPresentationRequestStatusReason.DebuggerPresented -> map["reason"] = "debuggerPresented"
    is PaywallPresentationRequestStatusReason.PaywallAlreadyPresented -> map["reason"] = "paywallAlreadyPresented"
    is PaywallPresentationRequestStatusReason.Holdout -> {
      map["reason"] = "holdout"
      map["experiment"] = Experiment.toJson(this.experiment)
    }
    is PaywallPresentationRequestStatusReason.NoAudienceMatch -> map["reason"] = "noAudienceMatch"
    is PaywallPresentationRequestStatusReason.PlacementNotFound -> map["reason"] = "placementNotFound"
    is PaywallPresentationRequestStatusReason.NoPaywallView -> map["reason"] = "noPaywallViewController"
    is PaywallPresentationRequestStatusReason.NoPresenter -> map["reason"] = "noPresenter"
    is PaywallPresentationRequestStatusReason.NoConfig -> map["reason"] = "noConfig"
    is PaywallPresentationRequestStatusReason.SubscriptionStatusTimeout -> map["reason"] = "subscriptionStatusTimeout"
  }
  return map
}
